flytekit/types/structured/structured_dataset.py

```diff
@@ -279,14 +299,17 @@ def to_literal(
         # StructuredDataset instance.
         if issubclass(python_type, StructuredDataset):
             assert isinstance(python_val, StructuredDataset)
-            df_type = type(python_val.dataframe)
+            python_val.uri = python_val.uri or ctx.file_access.get_random_remote_path()
+            protocol = protocol_prefix(python_val.uri)
             format = python_val.file_format
-            protocol = self.DEFAULT_PROTOCOLS[df_type]
-            try:
-                protocol = protocol_prefix(python_val.uri)
-            except ValueError:
-                ...
-
+            if protocol != StructuredDatasetProtocol.BIGQUERY:
+                format = python_val.file_format or StructuredDatasetFileFormat.PARQUET
+            if python_val.dataframe is None:
+                sd_model = literals.StructuredDataset(
+                    uri=python_val.uri, metadata=StructuredDatasetMetadata(format=format)
+                )
+                return Literal(scalar=Scalar(structured_dataset=sd_model))
```

Comment:
Can you explain why this is necessary?  This function should not return a Literal directly - that should always be redirected to one of the encoders.

Reply:
^